huaweicloud/services/iam/resource_huaweicloud_identity_provider_conversion.go

```diff
@@ -0,0 +1,316 @@
+	}
+	providerID := d.Get("provider_id").(string)
+	conversionID := MappingIDPrefix + providerID
+
+	// Check if the mappingID exists, update if it exists, otherwise create it.
+	r, err := mappings.List(client).AllPages()
+	err404 := golangsdk.ErrDefault404{}
+	if err != nil && !errors.As(err, &err404) {
+		return fmtp.DiagErrorf("error in querying or extract conversions: %s", err)
+	}
+
+	conversions, err := mappings.ExtractMappings(r)
+	if err != nil {
+		return fmtp.DiagErrorf("error in extracting provider conversions: %s", err)
+	}
+
+	filterData, err := utils.FilterSliceWithField(conversions, map[string]interface{}{
+		"ID": conversionID,
+	})
+	if err != nil {
+		return fmtp.DiagErrorf("error in filtering conversions: %s", err)
+	}
+
+	conversionRules := d.Get("conversion_rules").([]interface{})
+	mappingOpts := buildMappingRules(conversionRules)
+	// Create the mapping if it does not exist, otherwise update it.
+	if len(filterData) == 0 {
+		_, err = mappings.Create(client, conversionID, mappingOpts)
+	} else {
+		_, err = mappings.Update(client, conversionID, mappingOpts)
+	}
+	if err != nil {
+		return fmtp.DiagErrorf("error in creating/updating mapping: %s", err)
+	}
+	d.SetId(conversionID)
+
+	return resourceIAMProviderConversionRead(ctx, d, meta)
+}
+
+func buildMappingRules(conversionRules []interface{}) mappings.MappingOption {
+	rules := make([]mappings.MappingRule, 0, len(conversionRules))
+
+	for _, cr := range conversionRules {
+		convRule := cr.(map[string]interface{})
+
+		// build local rules
+		local := convRule["local"].([]interface{})
+		localRules := make([]mappings.LocalRule, 0, len(local))
+		for _, v := range local {
+			lRule := v.(map[string]interface{})
+			r := mappings.LocalRule{
+				User: mappings.LocalRuleVal{
+					Name: lRule["username"].(string),
+				},
+			}
+			group, ok := lRule["group"]
+			if ok && len(group.(string)) > 0 {
+				r.Group = &mappings.LocalRuleVal{
+					Name: group.(string),
+				}
+			}
+			localRules = append(localRules, r)
+		}
+		// build remote rule
+		remote := convRule["remote"].([]interface{})
+		remoteRules := make([]mappings.RemoteRule, 0, len(remote))
+		for _, v := range remote {
+			rRule := v.(map[string]interface{})
+			r := mappings.RemoteRule{
+				Type: rRule["attribute"].(string),
+			}
+			if condition, ok := rRule["condition"]; ok {
+				values := utils.ExpandToStringList(rRule["value"].([]interface{}))
+				if condition.(string) == "any_one_of" {
+					r.AnyOneOf = values
+				} else if condition.(string) == "not_any_of" {
+					r.NotAnyOf = values
+				}
+			}
+			remoteRules = append(remoteRules, r)
+		}
+
+		rule := mappings.MappingRule{
+			Local:  localRules,
+			Remote: remoteRules,
+		}
+		rules = append(rules, rule)
+	}
+	opts := mappings.MappingOption{
+		Rules: rules,
+	}
+	return opts
+}
+
+func buildConversionRules(conversionRules []interface{}) mappings.MappingOption {
```

Comment:
what's the difference between buildConversionRules  and buildMappingRules ?

Reply:
duplicate functions